components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java

```diff
@@ -706,47 +732,231 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
+                    if (log.isDebugEnabled()) {
+                        log.debug("DropUnregisteredScopes config is enabled. Attempting to drop unregistered scopes.");
+                    }
+                    allowedScopes = dropUnregisteredScopes(params);
                 }
-                allowedScopes = dropUnregisteredScopes(params);
+
+                // Get registered OIDC scopes.
+                String[] oidcScopes = oAuthAdminService.getScopeNames();
+                List<String> oidcScopeList = new ArrayList<>(Arrays.asList(oidcScopes));
+                for (String scope : allowedScopes) {
+                    if (!oidcScopeList.contains(scope)) {
+                        allowedOAuthScopes.add(scope);
+                    }
+                }
+            } catch (IdentityOAuthAdminException e) {
+                throw new OAuthSystemException("Error while retrieving OIDC scopes.", e);
+            } finally {
+                PrivilegedCarbonContext.endTenantFlow();
             }
+        }
+        return allowedOAuthScopes;
+    }
 
-            // Get registered OIDC scopes.
-            String[] oidcScopes = oAuthAdminService.getScopeNames();
-            List<String> oidcScopeList = new ArrayList<>(Arrays.asList(oidcScopes));
-
-            /* Remove OIDC scopes from scope string sent back
-             in the consent page as consent is not needed for
-             OIDC scopes.*/
-            StringBuilder scopes = new StringBuilder();
-            for (String scope : allowedScopes) {
-                if (!oidcScopeList.contains(scope)) {
-                    scopes.append(scope).append(" ");
+    private static String getConsentRequiredScopes(AuthenticatedUser user, OAuth2Parameters params)
+            throws OAuthSystemException {
+
+        try {
+            List<String> allowedOAuthScopes = getAllowedOAuthScopes(params);
+            if (user != null && !isPromptContainsConsent(params)) {
+                String userId = getUserIdOfAuthenticatedUser(user);
+                String appId = getAppIdFromClientId(params.getClientId());
+                OAuth2ScopeConsentResponse existingUserConsent = oAuth2ScopeService.getUserConsentForApp(
+                        userId, appId, IdentityTenantUtil.getTenantId(user.getTenantDomain()));
+                if (existingUserConsent != null) {
+                    if (CollectionUtils.isNotEmpty(existingUserConsent.getApprovedScopes())) {
+                        allowedOAuthScopes.removeAll(existingUserConsent.getApprovedScopes());
+                    }
                 }
             }
+            if (CollectionUtils.isNotEmpty(allowedOAuthScopes)) {
+                params.setConsentRequiredScopes(new HashSet<>(allowedOAuthScopes));
+                return allowedOAuthScopes.stream().collect(Collectors.joining(" ")).trim();
+            } else {
+                return StringUtils.EMPTY;
+            }
```

Comment:
Lets add a debug log to see the final set of scopes returned

Reply:
Fixed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1548/commits/27ddbc8fed9c4368ffe7a826b857537a6b8ffff0